src/main/scala/net/psforever/actors/session/SessionActor.scala

```diff
@@ -141,10002 +95,502 @@ object SessionActor {
-                    attemptRecoveryFromNoGuidExceptionAsPlayer(player, nge)
-                  case v: Vehicle =>
-                    if (v.PassengerInSeat(player).contains(0)) {
-                      attemptRecoveryFromNoGuidExceptionAsPlayer(player, nge)
-                    }
-                    SupervisorStrategy.resume
-                  case _ =>
-                    writeLogExceptionAndStop(nge)
-                }
-              case _ =>
-                //did not discover or resolve the situation
-                writeLogExceptionAndStop(nge)
-            }
-
-          case _ =>
-            SupervisorStrategy.resume
-        }
-
-      case ide: InventoryDisarrayException =>
-        attemptRecoveryFromInventoryDisarrayException(ide.inventory)
-        //re-evaluate results
-        if (ide.inventory.ElementsOnGridMatchList() > 0) {
-          writeLogExceptionAndStop(ide)
-        } else {
-          SupervisorStrategy.resume
-        }
-
-      case e =>
-        writeLogExceptionAndStop(e)
-    }
+  override def postStop(): Unit = {
+    //normally, the player avatar persists a minute or so after disconnect; we are subject to the SessionReaper
+    //TODO put any temporary values back into the avatar
+    sessionFuncs.stop()
   }
 
-  def attemptRecoveryFromNoGuidExceptionAsVehicle(v: Vehicle, e: Throwable): SupervisorStrategy.Directive = {
-    val entry = v.Seats.find { case (_, s) => s.occupants.contains(player) }
-    entry match {
-      case Some((index, _)) =>
-        player.VehicleSeated = None
-        v.Seats(0).unmount(player)
-        player.Position = v.Position
-        player.Orientation = v.Orientation
-        interstellarFerry = None
-        interstellarFerryTopLevelGUID = None
-        attemptRecoveryFromNoGuidExceptionAsPlayer(player, e)
-      case None =>
-        writeLogException(e)
-    }
-  }
+  def receive: Receive = startup
 
-  def attemptRecoveryFromNoGuidExceptionAsPlayer(p: Player, e: Throwable): SupervisorStrategy.Directive = {
-    if (p eq player) {
-      val hasGUID = p.HasGUID
-      zoneLoaded match {
-        case Some(true) if hasGUID =>
-          AvatarCreate() //this will probably work?
-          SupervisorStrategy.resume
-        case Some(false) =>
-          RequestSanctuaryZoneSpawn(p, continent.Number)
-          SupervisorStrategy.resume
-        case None =>
-          if (player.Zone eq Zone.Nowhere) {
-            RequestSanctuaryZoneSpawn(p, continent.Number)
-          } else {
-            zoneReload = true
-            LoadZoneAsPlayer(player, player.Zone.id)
-          }
-          SupervisorStrategy.resume
-        case _ =>
-          writeLogExceptionAndStop(e)
+  def startup: Receive = {
+    case msg =>
+      if (!sessionFuncs.assignEventBus(msg)) {
+        buffer.addOne(msg)
+      } else if (sessionFuncs.whenAllEventBusesLoaded()) {
+        context.become(inTheGame)
+        buffer.foreach { self.tell(_, self) }
+        buffer.clear()
```

Comment:
I can do that change, since this is new code, but this was one of those instances where I thought that breaking out the guard booleans actually makes it harder to understand what's going on.

Reply:
Changed.